packages/common/src/index-api.ts

```diff
@@ -28,19 +28,19 @@ export type Pagination = ForwardPagination | BackwardPagination
  * Boolean field value filter
  */
 export type BooleanValueFilter =
-  | { isNull: boolean }
-  | { equalTo: boolean }
-  | { notEqualTo: boolean }
+  | { type: 'bool'; op: 'null'; value: boolean }
```

Comment:
Since js-ceramic is a different system than js-composedb, this is the proposed public API for js-ceramic, since it does not utilized GraphQL. This is due to the separation of concerns that exists between the two systems

- js-ceramic: Types that provide information to construct queries for filtering, that can be easily represented in a REST request
- js-composedb: Graphql types that represent a query

The API of js-composedb will use types, that are represented as GraphQL for its inputs, while js-ceramic will use these types as its inputs. It is the responsibility of js-composedb to convert its inputs to inputs that js-ceramic requires.

It is possible there exists a type that requires no translation between the two systems, but the current implementation in js-ceramic uses the types in this PR, as they contain enough information to implement the necessary functionality, while being uniform enough to provide an understandable API for js-ceramic.

Please feel free to propose an alternate type for js-ceramic that meets its concerns.

Reply:
I think this is going backwards. What value does changing the types here bring, when we already discussed them in a previous PR (https://github.com/ceramicnetwork/js-ceramic/pull/2839), which was an explicit requirement to enable other work (https://github.com/ceramicstudio/js-composedb/pull/135)?
How does having different data shapes for filters in ComposeDB and Ceramic APIs serve our users in any way?

Reply:
The value here is that this is what is needed for js-ceramic to implement the required functionality. The previous PR said these are the filters that js-composedb would like to have, this PR builds on that and says this is the shape of those filters so the functionality can be implemented.

As far as serving users, this is the format for the Ceramic API that achieves the functionality users want, in a way that corresponds to its exposed interface. ComposeDB is not a concern of Ceramic, since the functionality that ComposeDB is implementing can be achieved with this API.

Again, if there is a format for these filters that exists to achieve this implementation without putting additional burden on users of the Ceramic API, please propose it.

Reply:
@dbcfd can you explain why these types are better than the existing types from the perspective of implementing the ceramic-side functionality?  They seem to be directly isomorphic to the existing types.  I agree that that there is a separation of concerns between Ceramic and Composedb, but if there's already a 1-1 mapping between the two types, why not just stick with the types that composedb uses for simplicity sake?

Also, I don't think that the logical "composedb" boundary ends at the beginning of the js-ceramic codebase, there is already a lot of composedb-specific code in js-ceramic.  I more and more already think of js-ceramic as the "composedb" server-side node.  But that's a bigger conversation for a different time.

Reply:
- Ease of use. For people using the api, this is a consistent interface for each query they input. For example, it's easier to set `{op: '=', value: 1}` than to remember `{equal: 1}` or is it `{eq: 1}`? And then if I need greater than, it's `{op='>', value: 1}`.
- Strictness of type: Without the type hint on the union, we may have to resort to any casting 
- Ease of implementation. Since this is a union field, we can build queries using the type, op, and value, rather than checking for individual fields to see if they are present. Although typescript might not allow that either.

This seemed like a good compromise between a stronger interface like codecs where we have to type each of the queries, and what was originally proposed as needed by composedb.

Reply:
> Ease of use. For people using the api, this is a consistent interface for each query they input. For example, it's easier to set {op: '=', value: 1} than to remember {equal: 1} or is it {eq: 1}? And then if I need greater than, it's {op: '>', value: 1}.

That seems pretty subjective.  This syntax is quite a bit more verbose, so I think some would say its less easy to use because of that.

> * Strictness of type: Without the type hint on the union, we may have to resort to any casting

Can't speak to that

> * Ease of implementation. Since this is a union field, we can build queries using the type, op, and value, rather than checking for individual fields to see if they are present. Although typescript might not allow that either.

This one I definitely *do* buy.  This format actually reads to me like an internal representation specifically designed for ease of implementation.  I could even see a world where the public API is the existing types used by graphql but then we translate that to a representation like this internally.


Overall, I ultimately don't feel *super* strongly about this one way or the other.  I do lean towards keeping the types the same between ComposeDB and Ceramic, as I think doing so may make things easier as we move to combine the codebases and move more composedb functionality server-side over time.  But if you and Paul can reach a consensus then I'm ultimately happy either way.

Reply:
As Spencer wrote, these APIs are designed for ComposeDB, it's not like there is another consumer for these we need to care about, and the different type structures are semantically the same.

I don't mind whatever internal representation is used, but it seems unnecessary to me. There's already a `convertQueryFilter()` that converts the API representation to the SQL representation so it's not like the suggested changes represent the canonical form of the filters for the DB anyways, it's just another abstraction.
I'd much rather have `GraphQL inputs as API structure representation` -> `SQL conversion` than `GraphQL inputs format` -> `intermediary API representation` -> `SQL conversion` where the intermediary representation doesn't bring any additional benefit, it's much easier to follow the logic, test and maintain.